File: android/PulsePhoneKey/app/src/main/java/com/teslapulse/phonekey/BlePairer.kt
package com.teslapulse.phonekey

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import java.util.UUID

class BlePairer(
    private val context: Context,
    private val onStatus: (String) -> Unit,
    private val onLog: (String) -> Unit,
    private val onWaitingCard: () -> Unit,
    private val onDone: () -> Unit,
) {
    private val handler = Handler(Looper.getMainLooper())
    private val adapter: BluetoothAdapter? =
        (context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager).adapter

    private var gatt: BluetoothGatt? = null
    private var writeChar: BluetoothGattCharacteristic? = null
    private var payload: ByteArray? = null
    private var targetNames: Set<String> = emptySet()
    private var scanning = false
    private var connected = false

    private val serviceUuid = UUID.fromString(VcsecPayload.SERVICE_UUID)
    private val writeUuid = UUID.fromString(VcsecPayload.WRITE_UUID)
    private val readUuid = UUID.fromString(VcsecPayload.READ_UUID)
    private val cccd = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

    @SuppressLint("MissingPermission")
    fun start(vin: String, publicKey: ByteArray) {
        stop()
        val names = VcsecPayload.bleNames(vin)
        targetNames = names.toSet()
        payload = VcsecPayload.addKeyRequest(publicKey)
        onLog("Hedef: ${names.joinToString(", ")}")
        onLog("Payload ${payload!!.size} byte")

        val bt = adapter
        if (bt == null || !bt.isEnabled) {
            onStatus("Bluetooth kapalı — Ayarlar’dan aç")
            return
        }

        onStatus("Tesla aranıyor…")
        scanning = true
        val scanner = bt.bluetoothLeScanner
        val filters = listOf(
            ScanFilter.Builder().setServiceUuid(ParcelUuid(serviceUuid)).build(),
        )
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()
        try {
            scanner.startScan(filters, settings, scanCallback)
        } catch (_: Exception) {
            scanner.startScan(scanCallback)
        }
        // Also unfiltered after 2s — some cars omit service UUID in adv
        handler.postDelayed({
            if (scanning && !connected) {
                try {
                    scanner.stopScan(scanCallback)
                    scanner.startScan(scanCallback)
                    onLog("Filtresiz tarama…")
                } catch (_: Exception) {
                }
            }
        }, 2000)
        handler.postDelayed({
            if (scanning && !connected) {
                stopScan()
                onStatus("Araç bulunamadı. Yakınlaş / arabayı uyandır.")
            }
        }, 28_000)
    }

    @SuppressLint("MissingPermission")
    fun stop() {
        stopScan()
        try {
            gatt?.close()
        } catch (_: Exception) {
        }
        gatt = null
        writeChar = null
        connected = false
    }

    @SuppressLint("MissingPermission")
    private fun stopScan() {
        if (!scanning) return
        scanning = false
        try {
            adapter?.bluetoothLeScanner?.stopScan(scanCallback)
        } catch (_: Exception) {
        }
    }

    private val scanCallback = object : ScanCallback() {
        @SuppressLint("MissingPermission")
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val name = result.device.name
                ?: result.scanRecord?.deviceName
                ?: ""
            val match = targetNames.contains(name)
                || name.startsWith("Tesla")
                || (name.startsWith("S") && name.endsWith("C") && name.length == 18)
            if (!match) return
            if (connected) return
            connected = true
            stopScan()
            onLog("Bulundu: $name (${result.rssi} dBm)")
            onStatus("Bağlanıyor: $name…")
            gatt = result.device.connectGatt(
                context,
                false,
                gattCallback,
                BluetoothDevice.TRANSPORT_LE,
            )
        }

        override fun onScanFailed(errorCode: Int) {
            onStatus("Tarama hatası: $errorCode")
            onLog("Scan failed $errorCode")
        }
    }

    private val gattCallback = object : BluetoothGattCallback() {
        @SuppressLint("MissingPermission")
        override fun onConnectionStateChange(g: BluetoothGatt, status: Int, newState: Int) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                onLog("GATT bağlı, servisler…")
                handler.post { onStatus("Servisler keşfediliyor…") }
                g.discoverServices()
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                handler.post { onLog("GATT koptu") }
            }
        }

        @SuppressLint("MissingPermission")
        override fun onServicesDiscovered(g: BluetoothGatt, status: Int) {
            val service = g.getService(serviceUuid)
            if (service == null) {
                handler.post { onStatus("Tesla servisi yok") }
                return
            }
            writeChar = service.getCharacteristic(writeUuid)
            val read = service.getCharacteristic(readUuid)
            if (writeChar == null) {
                handler.post { onStatus("Write characteristic yok") }
                return
            }
            if (read != null) {
                g.setCharacteristicNotification(read, true)
                val desc = read.getDescriptor(cccd)
                if (desc != null) {
                    if (Build.VERSION.SDK_INT >= 33) {
                        g.writeDescriptor(desc, BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE)
                    } else {
                        @Suppress("DEPRECATION")
                        desc.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                        @Suppress("DEPRECATION")
                        g.writeDescriptor(desc)
                    }
                }
            }
            handler.post {
                onStatus("add-key gönderiliyor…")
                writePayload(g)
            }
        }

        override fun onCharacteristicChanged(
            g: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            value: ByteArray,
        ) {
            handleNotify(value)
        }

        @Deprecated("Deprecated in Java")
        override fun onCharacteristicChanged(
            g: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
        ) {
            @Suppress("DEPRECATION")
            val value = characteristic.value ?: return
            handleNotify(value)
        }

        @SuppressLint("MissingPermission")
        override fun onCharacteristicWrite(
            g: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            status: Int,
        ) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                handler.post {
                    onWaitingCard()
                    onStatus("Key Card’ı KONSOLA koy → Pair / Confirm")
                    onLog("İstek gönderildi ✓")
                }
            } else {
                handler.post {
                    onStatus("Yazma hatası: $status")
                    onLog("Write failed $status")
                }
            }
        }
    }

    @SuppressLint("MissingPermission")
    private fun writePayload(g: BluetoothGatt) {
        val data = payload ?: return
        val ch = writeChar ?: return
        // Single write is fine for ~88 bytes on modern phones; chunk if needed
        if (Build.VERSION.SDK_INT >= 33) {
            g.writeCharacteristic(ch, data, BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT)
        } else {
            @Suppress("DEPRECATION")
            ch.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
            @Suppress("DEPRECATION")
            ch.value = data
            @Suppress("DEPRECATION")
            g.writeCharacteristic(ch)
        }
    }

    private fun handleNotify(value: ByteArray) {
        val hex = value.joinToString("") { "%02x".format(it) }
        handler.post {
            onLog("RX ${hex.take(48)}…")
            if (hex.contains("0801") || hex.contains("2202")) {
                onWaitingCard()
                onStatus("Araç kart bekliyor — konsola Key Card koy")
            }
            if (hex.contains("1a08") || hex.contains("5f0d")) {
                onDone()
                onStatus("Onaylandı — Phone Key eklendi")
                onLog("Whitelist OK (muhtemel)")
            }
        }
    }
}

File: ios/PulsePhoneKey/Sources/BLEPairer.swift
import Foundation
import CoreBluetooth
import Combine

/// Robust Tesla VCSEC BLE pairer for iPad / Swift Playgrounds.
///
/// Critical: writes use `.withResponse` and wait for each ACK before the next
/// chunk. Flooding writes (sleep-only) drops bytes → car never shows Pair.
@MainActor
final class BLEPairer: NSObject, ObservableObject {
    @Published var status: String = "Hazır — önce arabayı uyandır"
    @Published var log: [String] = []
    @Published var busy = false
    @Published var waitingForCard = false
    @Published var paired = false
    @Published var lastDetail: String = ""

    private var central: CBCentralManager!
    private var targetNames: Set<String> = []
    private var payload: Data?
    private var peripheral: CBPeripheral?
    private var writeChar: CBCharacteristic?
    private var readChar: CBCharacteristic?

    private var connectContinuation: CheckedContinuation<Void, Error>?
    private var writeContinuation: CheckedContinuation<Void, Error>?
    private var notifyContinuation: CheckedContinuation<Void, Error>?
    private var scanTimeoutTask: Task<Void, Never>?
    private var connecting = false
    private var wrotePayload = false
    private var reconnectAttempts = 0

    override init() {
        super.init()
        central = CBCentralManager(delegate: self, queue: nil, options: [
            CBCentralManagerOptionShowPowerAlertKey: true,
        ])
    }

    func appendLog(_ line: String) {
        let stamped = "\(Self.clock()) \(line)"
        log.insert(stamped, at: 0)
        if log.count > 60 { log = Array(log.prefix(60)) }
    }

    func pair(vin: String, publicKey: Data) async {
        busy = true
        paired = false
        waitingForCard = false
        wrotePayload = false
        connecting = false
        reconnectAttempts = 0
        writeChar = nil
        readChar = nil
        defer { busy = false }

        resetContinuations(with: PairError.cancelled)

        let names = VCSECPayload.bleNames(vin: vin)
        targetNames = Set(names)
        payload = VCSECPayload.addKeyRequest(publicKeyUncompressed: publicKey)
        appendLog("Hedef: \(names.joined(separator: ", "))")
        appendLog("Payload \(payload!.count) byte")
        lastDetail = "S…C listede kaybolursa normal — bağlanınca gizlenir."

        guard central.state == .poweredOn else {
            status = "Bluetooth kapalı — Ayarlar’dan aç"
            appendLog("Bluetooth state: \(central.state.rawValue)")
            return
        }

        // Wake tip first — advertising stops when the car sleeps.
        status = "Arabayı uyandır (kapı/ekran) → Tesla aranıyor…"
        do {
            try await scanAndConnect(timeout: 45)
            status = "Bildirim açılıyor…"
            try await enableNotify()
            status = "add-key gönderiliyor…"
            try await writePayload()
            wrotePayload = true
            waitingForCard = true
            status = "✓ İstek gitti — Key Card’ı KONSOLA koy → Pair"
            lastDetail = "Uygulamadan çıkma. Kartı iPad’e değil konsola koy."
            appendLog("İstek gönderildi ✓ — kartı konsola koy")
            // Stay connected so the car can reply WAIT / OK.
            try? await Task.sleep(nanoseconds: 90_000_000_000)
        } catch {
            status = "Hata: \(error.localizedDescription)"
            lastDetail = Self.hint(for: error)
            appendLog(error.localizedDescription)
        }
    }

    // MARK: - Scan / connect

    private func scanAndConnect(timeout: TimeInterval) async throws {
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            self.connectContinuation = cont
            startScan()
            scanTimeoutTask = Task { [weak self] in
                try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                await MainActor.run {
                    guard let self, let c = self.connectContinuation else { return }
                    self.connectContinuation = nil
                    self.central.stopScan()
                    c.resume(throwing: PairError.timeout)
                }
            }
        }
    }

    private func startScan() {
        central.stopScan()
        let service = CBUUID(string: VCSECPayload.serviceUUID)
        appendLog("Tarama (servis filtresi)…")
        central.scanForPeripherals(withServices: [service], options: [
            CBCentralManagerScanOptionAllowDuplicatesKey: true,
        ])
        // Many cars omit service UUID in advertisement — open scan after 2s.
        Task { [weak self] in
            try? await Task.sleep(nanoseconds: 2_000_000_000)
            await MainActor.run {
                guard let self, self.connectContinuation != nil, !self.connecting else { return }
                self.appendLog("Filtresiz tarama…")
                self.central.stopScan()
                self.central.scanForPeripherals(withServices: nil, options: [
                    CBCentralManagerScanOptionAllowDuplicatesKey: true,
                ])
            }
        }
    }

    private func finishConnect(success: Bool, error: Error? = nil) {
        scanTimeoutTask?.cancel()
        scanTimeoutTask = nil
        central.stopScan()
        guard let c = connectContinuation else { return }
        connectContinuation = nil
        if success {
            c.resume()
        } else {
            c.resume(throwing: error ?? PairError.notReady)
        }
    }

    // MARK: - Notify + write

    private func enableNotify() async throws {
        guard let peripheral, let readChar else {
            // Notify is helpful but not strictly required for add-key.
            appendLog("Read char yok — yine de yazılacak")
            return
        }
        if readChar.isNotifying { return }
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            self.notifyContinuation = cont
            peripheral.setNotifyValue(true, for: readChar)
            Task { [weak self] in
                try? await Task.sleep(nanoseconds: 3_000_000_000)
                await MainActor.run {
                    guard let self, let n = self.notifyContinuation else { return }
                    self.notifyContinuation = nil
                    // Proceed even if notify ACK is slow.
                    n.resume()
                }
            }
        }
    }

    private func writePayload() async throws {
        guard let peripheral, let writeChar, let payload else {
            throw PairError.notReady
        }
        // Prefer one shot when MTU allows (Android path). Else chunk with ACK.
        let mtu = max(20, peripheral.maximumWriteValueLength(for: .withResponse))
        appendLog("MTU yazma: \(mtu) byte")
        var offset = 0
        var part = 0
        while offset < payload.count {
            let end = min(offset + mtu, payload.count)
            let chunk = payload.subdata(in: offset..<end)
            part += 1
            appendLog("TX \(part) \(chunk.count) byte…")
            try await writeChunk(chunk, peripheral: peripheral, characteristic: writeChar)
            offset = end
        }
        appendLog("TX tamam (\(payload.count) byte)")
    }

    private func writeChunk(
        _ data: Data,
        peripheral: CBPeripheral,
        characteristic: CBCharacteristic
    ) async throws {
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            self.writeContinuation = cont
            peripheral.writeValue(data, for: characteristic, type: .withResponse)
            Task { [weak self] in
                try? await Task.sleep(nanoseconds: 8_000_000_000)
                await MainActor.run {
                    guard let self, let w = self.writeContinuation else { return }
                    self.writeContinuation = nil
                    w.resume(throwing: PairError.writeTimeout)
                }
            }
        }
    }

    private func resetContinuations(with error: Error) {
        connectContinuation?.resume(throwing: error)
        connectContinuation = nil
        writeContinuation?.resume(throwing: error)
        writeContinuation = nil
        notifyContinuation?.resume(throwing: error)
        notifyContinuation = nil
        scanTimeoutTask?.cancel()
        scanTimeoutTask = nil
        central?.stopScan()
    }

    private static func clock() -> String {
        let f = DateFormatter()
        f.dateFormat = "HH:mm:ss"
        return f.string(from: Date())
    }

    private static func hint(for error: Error) -> String {
        if let e = error as? PairError {
            switch e {
            case .timeout:
                return "Kapıyı aç / ekranı uyandır. Tesla uygulamasını kapat. iPad’i direksiyona yakın tut."
            case .disconnected:
                return "Bağlantı koptu. Arabayı uyandırıp tekrar dene. Ayarlar’da S…C kaybolması bağlanınca normal."
            case .writeTimeout, .writeFailed:
                return "Yazma tamamlanmadı. Playgrounds’ta Run ▶ ile çalıştır; uygulamadan çıkma."
            default:
                return "Tekrar dene. Log’a bak."
            }
        }
        return error.localizedDescription
    }

    enum PairError: LocalizedError {
        case timeout, notReady, bluetoothOff, disconnected, writeTimeout, writeFailed, cancelled
        var errorDescription: String? {
            switch self {
            case .timeout:
                return "Araç bulunamadı (BLE). Uyandır / yakınlaş."
            case .notReady:
                return "BLE hazır değil (servis/karakteristik)."
            case .bluetoothOff:
                return "Bluetooth kapalı"
            case .disconnected:
                return "GATT bağlantısı koptu"
            case .writeTimeout:
                return "Yazma zaman aşımı"
            case .writeFailed:
                return "Yazma başarısız"
            case .cancelled:
                return "İptal"
            }
        }
    }
}

extension BLEPairer: CBCentralManagerDelegate, CBPeripheralDelegate {
    nonisolated func centralManagerDidUpdateState(_ central: CBCentralManager) {
        Task { @MainActor in
            switch central.state {
            case .poweredOn: appendLog("Bluetooth açık")
            case .unauthorized: status = "Bluetooth izni yok — Ayarlar → Playgrounds"
            case .poweredOff: status = "Bluetooth kapalı"
            default: appendLog("Bluetooth state \(central.state.rawValue)")
            }
        }
    }

    nonisolated func centralManager(
        _ central: CBCentralManager,
        didDiscover peripheral: CBPeripheral,
        advertisementData: [String: Any],
        rssi RSSI: NSNumber
    ) {
        let name = peripheral.name
            ?? (advertisementData[CBAdvertisementDataLocalNameKey] as? String)
            ?? ""
        Task { @MainActor in
            guard connectContinuation != nil, !connecting else { return }
            let match = targetNames.contains(name)
                || name.hasPrefix("Tesla")
                || (name.hasPrefix("S") && name.hasSuffix("C") && name.count == 18)
            guard match else { return }

            connecting = true
            appendLog("Bulundu: \(name) (\(RSSI) dBm)")
            lastDetail = "\(name) bulundu — Ayarlar’dan kaybolması normal."
            self.central.stopScan()
            self.peripheral = peripheral
            peripheral.delegate = self
            status = "Bağlanıyor: \(name)…"
            self.central.connect(peripheral, options: [
                CBConnectPeripheralOptionNotifyOnConnectionKey: true,
                CBConnectPeripheralOptionNotifyOnDisconnectionKey: true,
            ])
        }
    }

    nonisolated func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        Task { @MainActor in
            appendLog("GATT bağlı — servisler…")
            status = "Servisler keşfediliyor…"
            peripheral.discoverServices([CBUUID(string: VCSECPayload.serviceUUID)])
        }
    }

    nonisolated func centralManager(
        _ central: CBCentralManager,
        didFailToConnect peripheral: CBPeripheral,
        error: Error?
    ) {
        Task { @MainActor in
            connecting = false
            appendLog("Bağlantı başarısız: \(error?.localizedDescription ?? "?")")
            finishConnect(success: false, error: error ?? PairError.notReady)
        }
    }

    nonisolated func centralManager(
        _ central: CBCentralManager,
        didDisconnectPeripheral peripheral: CBPeripheral,
        error: Error?
    ) {
        Task { @MainActor in
            appendLog("GATT koptu: \(error?.localizedDescription ?? "ok")")
            if wrotePayload {
                // After add-key, some cars drop the link; Pair UI can still appear.
                if waitingForCard {
                    status = "Bağlantı koptu — yine de Key Card’ı konsola dene"
                }
                return
            }
            if connectContinuation != nil {
                // Still in handshake — one automatic reconnect.
                if reconnectAttempts < 2 {
                    reconnectAttempts += 1
                    connecting = true
                    appendLog("Yeniden bağlanılıyor (\(reconnectAttempts))…")
                    status = "Kopuk — yeniden bağlanılıyor…"
                    central.connect(peripheral, options: nil)
                    return
                }
                finishConnect(success: false, error: PairError.disconnected)
                return
            }
            if writeContinuation != nil {
                let w = writeContinuation
                writeContinuation = nil
                w?.resume(throwing: PairError.disconnected)
            }
        }
    }

    nonisolated func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        Task { @MainActor in
            if let error {
                finishConnect(success: false, error: error)
                return
            }
            guard let service = peripheral.services?.first(where: {
                $0.uuid == CBUUID(string: VCSECPayload.serviceUUID)
            }) else {
                appendLog("Tesla servisi yok")
                finishConnect(success: false, error: PairError.notReady)
                return
            }
            appendLog("Servis OK — karakteristikler…")
            peripheral.discoverCharacteristics(
                [
                    CBUUID(string: VCSECPayload.writeUUID),
                    CBUUID(string: VCSECPayload.readUUID),
                ],
                for: service
            )
        }
    }

    nonisolated func peripheral(
        _ peripheral: CBPeripheral,
        didDiscoverCharacteristicsFor service: CBService,
        error: Error?
    ) {
        Task { @MainActor in
            if let error {
                finishConnect(success: false, error: error)
                return
            }
            writeChar = service.characteristics?.first {
                $0.uuid == CBUUID(string: VCSECPayload.writeUUID)
            }
            readChar = service.characteristics?.first {
                $0.uuid == CBUUID(string: VCSECPayload.readUUID)
            }
            if writeChar != nil {
                appendLog("Write/Read karakteristik hazır")
                finishConnect(success: true)
            } else {
                appendLog("Write characteristic yok")
                finishConnect(success: false, error: PairError.notReady)
            }
        }
    }

    nonisolated func peripheral(
        _ peripheral: CBPeripheral,
        didUpdateNotificationStateFor characteristic: CBCharacteristic,
        error: Error?
    ) {
        Task { @MainActor in
            if let error {
                appendLog("Notify hata: \(error.localizedDescription)")
            } else {
                appendLog("Notify \(characteristic.isNotifying ? "açık" : "kapalı")")
            }
            notifyContinuation?.resume()
            notifyContinuation = nil
        }
    }

    nonisolated func peripheral(
        _ peripheral: CBPeripheral,
        didWriteValueFor characteristic: CBCharacteristic,
        error: Error?
    ) {
        Task { @MainActor in
            guard let w = writeContinuation else { return }
            writeContinuation = nil
            if let error {
                appendLog("Write hata: \(error.localizedDescription)")
                w.resume(throwing: error)
            } else {
                appendLog("Write ACK ✓")
                w.resume()
            }
        }
    }

    nonisolated func peripheral(
        _ peripheral: CBPeripheral,
        didUpdateValueFor characteristic: CBCharacteristic,
        error: Error?
    ) {
        guard let data = characteristic.value else { return }
        let hex = data.map { String(format: "%02x", $0) }.joined()
        Task { @MainActor in
            appendLog("RX \(hex.prefix(64))")
            // WAIT / card present signals commonly seen on VCSEC
            if hex.contains("0801") || hex.contains("2202") {
                waitingForCard = true
                status = "Araç kart bekliyor — konsola Key Card koy"
                lastDetail = "Şimdi kartı konsol okuyucuya koy; ekranda Pair çıkmalı."
            }
            if hex.contains("1a08") || hex.contains("5f0d") {
                paired = true
                status = "Onaylandı — Phone Key eklendi"
                appendLog("Whitelist OK (muhtemel)")
            }
        }
    }
}
